### PR TITLE
[SPARK-18493] Add missing python APIs: withWatermark and checkpoint to dataframe

### PR DESCRIPTION
**Pull request by brkyvz**
## What changes were proposed in this pull request?

This PR adds two of the newly added methods of `Dataset`s to Python:
`withWatermark` and `checkpoint`

## How was this patch tested?

Doc tests

### CODE REVIEW

**Comment by brkyvz** [comment]
cc @davies for PySpark changes
cc @liancheng for `checkpoint` API and javadoc update
cc @marmbrus for `withWatermark` API. Question here: should we throw an analysis exception if the Dataset that `withWatermark` is called on is non-streaming?

**Comment by marmbrus** [comment]
No, I don't think we need to throw any exceptions.  Watermarks are defined at batch boundaries, so it would just have no affect for a batch job.

We should make sure that the batch planner knows that it can elide the operator (it probably does not today).

**Comment by rxin** [comment]
@marmbrus Why don't we want to throw exceptions? Wouldn't it help users catch errors early.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -485,7 +485,10 @@ class Dataset[T] private[sql](
   def isStreaming: Boolean = logicalPlan.isStreaming
 
   /**
-   * Returns a checkpointed version of this Dataset.
+   * Eagerly checkpoint a Dataset and return the new Dataset. Checkpointing can be used to truncate
+   * the logical plan of this Dataset, which is especially useful in iterative algorithms where the
+   * plan may grow exponentially. It will be saved to a file inside the checkpoint
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nit: `a file` -> `files`.

Each partition will be saved to one file.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -495,7 +498,10 @@ class Dataset[T] private[sql](
   def checkpoint(): Dataset[T] = checkpoint(eager = true)
 
   /**
-   * Returns a checkpointed version of this Dataset.
+   * Returns a checkpointed version of this Dataset. Checkpointing can be used to truncate the
+   * logical plan of this Dataset, which is especially useful in iterative algorithms where the
+   * plan may grow exponentially. It will be saved to a file inside the checkpoint
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nit: `a file` -> `files`.

Each partition will be saved to one file.

File: python/pyspark/sql/dataframe.py [hunk 3]
```diff
@@ -322,6 +322,53 @@ def show(self, n=20, truncate=True):
     def __repr__(self):
         return "DataFrame[%s]" % (", ".join("%s: %s" % c for c in self.dtypes))
 
+    @since(2.1)
+    def checkpoint(self, eager=True):
+        """Returns a checkpointed version of this Dataset. Checkpointing can be used to truncate the
+        logical plan of this DataFrame, which is especially useful in iterative algorithms where the
+        plan may grow exponentially. It will be saved to a file inside the checkpoint
```
Comment (zsxwing) [inline python/pyspark/sql/dataframe.py]: nit: `a file` -> `files`.

Each partition will be saved to one file.

**Review by zsxwing — CHANGES_REQUESTED** [review]
Looks good overall. Just several nits.

**Comment by brkyvz** [comment]
thanks @gatorsmile and @tdas. I addressed your comments. The semantics look a lot cleaner now. That doesn't still mean it's clean though :P

**Comment by brkyvz** [comment]
Hallelujah! @zsxwing shall we merge this?

**Comment by zsxwing** [comment]
LGMT. Merging to master and 2.1.

**Pull request closed without merging**